packages/protocol/contracts/ACE/validators/privateRange/PrivateRange.sol

```diff
@@ -0,0 +1,346 @@
+                    }
+
+                    // Calculate the G1 element \gamma_i^{k}h^{a}\sigma_i^{-c} = B_i
+                    // Memory map:
+                    // 0x20: \gamma_iX
+                    // 0x40: \gamma_iY
+                    // 0x60: k_i
+                    // 0x80: hX
+                    // 0xa0: hY
+                    // 0xc0: a_i
+                    // 0xe0: \sigma_iX
+                    // 0x100: \sigma_iY
+                    // 0x120: -c
+                    calldatacopy(0xe0, add(noteIndex, 0x80), 0x40)
+                    calldatacopy(0x20, add(noteIndex, 0x40), 0x40)
+                    mstore(0x120, sub(gen_order, c)) 
+                    mstore(0x60, k)
+                    mstore(0xc0, a)
+
+                    // Call bn128 scalar multiplication precompiles
+                    // Represent point + multiplication scalar in 3 consecutive blocks of memory
+                    // Store \sigma_i^{-c} at 0x1a0:0x200
+                    // Store \gamma_i^{k} at 0x120:0x160
+                    // Store h^{a} at 0x160:0x1a0
+                    let result := staticcall(gas, 7, 0xe0, 0x60, 0x1a0, 0x40)
+                    result := and(result, staticcall(gas, 7, 0x20, 0x60, 0x120, 0x40))
+                    result := and(result, staticcall(gas, 7, 0x80, 0x60, 0x160, 0x40))
+
+                    // Call bn128 group addition precompiles
+                    // \gamma_i^{k} and h^{a} in memory block 0x120:0x1a0
+                    // Store result of addition at 0x160:0x1a0
+                    result := and(result, staticcall(gas, 6, 0x120, 0x80, 0x160, 0x40))
+
+                    // \gamma_i^{k}h^{a} and \sigma^{-c} in memory block 0x160:0x1e0
+                    // Store resulting point B at memory index b
+                    result := and(result, staticcall(gas, 6, 0x160, 0x80, b, 0x40))
+                    
+                    if eq(i, 0) { // m = 0
+                        mstore(0x260, mload(0x20))
+                        mstore(0x280, mload(0x40))
+                        mstore(0x1e0, mload(0xe0))
+                        mstore(
+                            0x200,
+                            sub(0x30644e72e131a029b85045b68181585d97816a916871ca8d3c208c16d87cfd47, mload(0x100))
+                            )
+                    }
+
+                    // If i > m + 1 (i.e. subsequent output notes)
+                    // then we add \sigma^{-c} and \sigma_{acc} and store result at \sigma_{acc} (0x1e0:0x200)
+                    // we then calculate \gamma^{cx} and add into \gamma_{acc}
+                    if gt(i, 0) {  // performing bilinear pairing comparison on all notes
+                        mstore(0x60, c)
+
+                        result := and(
+                            result,
+                            and(
+                                and(
+                                    staticcall(gas, 6, 0x1a0, 0x80, 0x1e0, 0x40),
```

Comment:
We have 3 notes in this proof, but we are only checking the pairing relationship on 2 notes

In this section, we need to update the memory location that we reference when we call the ecAdd precompile, so that we store the 3 notes linearly in memory